File: app/services/backup_service.py
"""Backup and Restore module (PRD section 12).

Plain JSON/ZIP export-import plus a password-protected AES-256-GCM encrypted
backup format. Note: this encrypts *backup files*, not the live SQLite
database at rest (that would require SQLCipher, which has no prebuilt wheel
for this Python/Windows/Android combination without a custom native build —
out of scope here). PIN/password app-lock (see settings_repository) covers
access control for the live app.
"""
from __future__ import annotations

import json
import os
import shutil
import zipfile
from datetime import datetime
from pathlib import Path
from typing import Optional

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

from app.core.database import get_db

# Parents before children, so a restore can insert in this order with
# foreign_keys temporarily disabled.
TABLE_ORDER = [
    "schools", "children", "academic_years", "classes", "terms", "subjects",
    "exams", "marks", "teachers", "teacher_assignments", "expense_categories",
    "expenses", "fee_receipts", "attachments", "ocr_history", "predictions",
    "backups", "settings",
]

ATTACHMENT_FIELDS = {
    "children": ["photo_path"],
    "expenses": ["receipt_path"],
    "fee_receipts": ["image_path"],
}

MAGIC = b"EDU1"
_PBKDF2_ITERATIONS = 200_000


def _app_data_dir() -> Path:
    try:
        from kivy.app import App

        running = App.get_running_app()
        if running is not None:
            return Path(running.user_data_dir)
    except Exception:
        pass
    return Path(__file__).resolve().parent.parent / "data"


def backups_dir() -> Path:
    path = _app_data_dir() / "backups"
    path.mkdir(parents=True, exist_ok=True)
    return path


def attachments_restore_dir() -> Path:
    path = _app_data_dir() / "attachments"
    path.mkdir(parents=True, exist_ok=True)
    return path


def _timestamped_name(prefix: str, ext: str) -> str:
    return f"{prefix}_{datetime.now().strftime('%Y%m%d_%H%M%S')}.{ext}"


def record_backup(path: Path, backup_type: str, conn=None) -> None:
    conn = conn or get_db()
    conn.execute(
        "INSERT INTO backups (file_path, backup_type) VALUES (?, ?)", (str(path), backup_type)
    )
    conn.commit()


def last_backup(conn=None):
    conn = conn or get_db()
    return conn.execute("SELECT * FROM backups ORDER BY created_at DESC LIMIT 1").fetchone()


def export_all(conn=None) -> dict:
    conn = conn or get_db()
    data = {}
    for table in TABLE_ORDER:
        rows = conn.execute(f"SELECT * FROM {table}").fetchall()
        data[table] = [dict(r) for r in rows]
    return data


def restore_from_dict(data: dict, conn=None) -> None:
    conn = conn or get_db()
    conn.execute("PRAGMA foreign_keys = OFF")
    try:
        for table in reversed(TABLE_ORDER):
            conn.execute(f"DELETE FROM {table}")
        for table in TABLE_ORDER:
            for row in data.get(table, []):
                columns = list(row.keys())
                placeholders = ", ".join("?" for _ in columns)
                conn.execute(
                    f"INSERT INTO {table} ({', '.join(columns)}) VALUES ({placeholders})",
                    [row[c] for c in columns],
                )
        conn.commit()
    except Exception:
        conn.rollback()
        raise
    finally:
        conn.execute("PRAGMA foreign_keys = ON")


# --- Plain JSON ---
def export_json(path: Optional[Path] = None, conn=None) -> Path:
    path = path or backups_dir() / _timestamped_name("backup", "json")
    with open(path, "w", encoding="utf-8") as fh:
        json.dump(export_all(conn), fh, indent=2, default=str)
    record_backup(path, "json", conn=conn)
    return path


def import_json(path: Path, conn=None) -> None:
    with open(path, "r", encoding="utf-8") as fh:
        data = json.load(fh)
    restore_from_dict(data, conn=conn)


# --- ZIP (JSON + attachment files) ---
def export_zip(path: Optional[Path] = None, conn=None) -> Path:
    data = export_all(conn)
    path = path or backups_dir() / _timestamped_name("backup", "zip")
    with zipfile.ZipFile(path, "w", zipfile.ZIP_DEFLATED) as zf:
        for table, fields in ATTACHMENT_FIELDS.items():
            for row in data.get(table, []):
                for field in fields:
                    src = row.get(field)
                    if src and Path(src).exists():
                        arcname = f"files/{Path(src).name}"
                        zf.write(src, arcname)
                        row[field] = arcname
        zf.writestr("data.json", json.dumps(data, indent=2, default=str))
    record_backup(path, "zip", conn=conn)
    return path


def import_zip(path: Path, conn=None) -> None:
    extract_dir = attachments_restore_dir()
    with zipfile.ZipFile(path, "r") as zf:
        data = json.loads(zf.read("data.json").decode("utf-8"))
        for table, fields in ATTACHMENT_FIELDS.items():
            for row in data.get(table, []):
                for field in fields:
                    rel = row.get(field)
                    if rel and str(rel).startswith("files/"):
                        target = extract_dir / Path(rel).name
                        with zf.open(rel) as src, open(target, "wb") as dst:
                            shutil.copyfileobj(src, dst)
                        row[field] = str(target)
    restore_from_dict(data, conn=conn)


# --- AES-256-GCM encrypted backup ---
def _derive_key(password: str, salt: bytes) -> bytes:
    kdf = PBKDF2HMAC(algorithm=hashes.SHA256(), length=32, salt=salt, iterations=_PBKDF2_ITERATIONS)
    return kdf.derive(password.encode("utf-8"))


def export_encrypted(password: str, path: Optional[Path] = None, conn=None) -> Path:
    payload = json.dumps(export_all(conn), default=str).encode("utf-8")
    salt = os.urandom(16)
    nonce = os.urandom(12)
    key = _derive_key(password, salt)
    ciphertext = AESGCM(key).encrypt(nonce, payload, None)

    path = path or backups_dir() / _timestamped_name("backup_encrypted", "bak")
    with open(path, "wb") as fh:
        fh.write(MAGIC + salt + nonce + ciphertext)
    record_backup(path, "encrypted", conn=conn)
    return path


def import_encrypted(password: str, path: Path, conn=None) -> None:
    raw = Path(path).read_bytes()
    if raw[:4] != MAGIC:
        raise ValueError("Not a recognized encrypted backup file")
    salt, nonce, ciphertext = raw[4:20], raw[20:32], raw[32:]
    key = _derive_key(password, salt)
    try:
        payload = AESGCM(key).decrypt(nonce, ciphertext, None)
    except Exception as exc:
        raise ValueError("Incorrect password or corrupted backup file") from exc
    restore_from_dict(json.loads(payload.decode("utf-8")), conn=conn)

File: android-native/app/src/main/java/org/familytools/educationtracker/services/BackupService.kt
package org.familytools.educationtracker.services

import android.content.Context
import androidx.sqlite.db.SupportSQLiteDatabase
import org.familytools.educationtracker.data.AppDatabase
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.security.SecureRandom
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Backup/Restore module. Unlike the Kivy build (where the `cryptography`
 * package needed a Rust toolchain to cross-compile for Android and had to
 * be made optional), AES-256-GCM here uses javax.crypto — part of the
 * Android platform, always available, no extra dependency at all.
 */
object BackupService {
    // Parents before children, so a restore can insert in this order with
    // foreign keys off.
    private val TABLE_ORDER = listOf(
        "children", "academic_years", "classes", "terms", "subjects", "exams", "marks",
        "teachers", "teacher_assignments", "expense_categories", "expenses", "fee_receipts",
        "ocr_history", "backups",
    )

    private const val MAGIC = "EDU1"
    private const val PBKDF2_ITERATIONS = 200_000

    private fun db(context: Context): SupportSQLiteDatabase =
        AppDatabase.getInstance(context).openHelper.writableDatabase

    private fun exportAllToJson(context: Context): JSONObject {
        val db = db(context)
        val root = JSONObject()
        for (table in TABLE_ORDER) {
            val array = JSONArray()
            db.query("SELECT * FROM $table").use { cursor ->
                while (cursor.moveToNext()) {
                    val row = JSONObject()
                    for (i in 0 until cursor.columnCount) {
                        val name = cursor.getColumnName(i)
                        when (cursor.getType(i)) {
                            android.database.Cursor.FIELD_TYPE_NULL -> row.put(name, JSONObject.NULL)
                            android.database.Cursor.FIELD_TYPE_INTEGER -> row.put(name, cursor.getLong(i))
                            android.database.Cursor.FIELD_TYPE_FLOAT -> row.put(name, cursor.getDouble(i))
                            else -> row.put(name, cursor.getString(i))
                        }
                    }
                    array.put(row)
                }
            }
            root.put(table, array)
        }
        return root
    }

    private fun restoreFromJson(context: Context, json: JSONObject) {
        val db = db(context)
        db.execSQL("PRAGMA foreign_keys = OFF")
        db.beginTransaction()
        try {
            for (table in TABLE_ORDER.reversed()) db.execSQL("DELETE FROM $table")
            for (table in TABLE_ORDER) {
                val array = json.optJSONArray(table) ?: continue
                for (i in 0 until array.length()) {
                    val row = array.getJSONObject(i)
                    val columns = row.keys().asSequence().toList()
                    val placeholders = columns.joinToString(",") { "?" }
                    val values = columns.map { row.get(it).takeIf { v -> v != JSONObject.NULL } }.toTypedArray()
                    db.execSQL(
                        "INSERT INTO $table (${columns.joinToString(",")}) VALUES ($placeholders)",
                        values,
                    )
                }
            }
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
            db.execSQL("PRAGMA foreign_keys = ON")
        }
    }

    private fun timestampedName(prefix: String, ext: String): String {
        val ts = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(Date())
        return "${prefix}_$ts.$ext"
    }

    private fun backupsDir(context: Context): File =
        File(context.getExternalFilesDir(null), "backups").apply { mkdirs() }

    private fun recordBackup(context: Context, path: String, type: String) {
        db(context).execSQL(
            "INSERT INTO backups (filePath, backupType, createdAt) VALUES (?, ?, ?)",
            arrayOf(path, type, System.currentTimeMillis()),
        )
    }

    fun lastBackupPath(context: Context): Pair<String, Long>? {
        db(context).query("SELECT filePath, createdAt FROM backups ORDER BY createdAt DESC LIMIT 1").use {
            return if (it.moveToFirst()) it.getString(0) to it.getLong(1) else null
        }
    }

    fun exportJson(context: Context): File {
        val file = File(backupsDir(context), timestampedName("backup", "json"))
        file.writeText(exportAllToJson(context).toString(2))
        recordBackup(context, file.absolutePath, "json")
        return file
    }

    fun importJson(context: Context, file: File) {
        restoreFromJson(context, JSONObject(file.readText()))
    }

    fun exportZip(context: Context): File {
        val json = exportAllToJson(context)
        val file = File(backupsDir(context), timestampedName("backup", "zip"))
        ZipOutputStream(file.outputStream()).use { zip ->
            zip.putNextEntry(ZipEntry("data.json"))
            zip.write(json.toString(2).toByteArray())
            zip.closeEntry()
        }
        recordBackup(context, file.absolutePath, "zip")
        return file
    }

    fun importZip(context: Context, file: File) {
        ZipInputStream(file.inputStream()).use { zip ->
            var entry = zip.nextEntry
            while (entry != null) {
                if (entry.name == "data.json") {
                    val text = zip.readBytes().toString(Charsets.UTF_8)
                    restoreFromJson(context, JSONObject(text))
                    return
                }
                entry = zip.nextEntry
            }
        }
        throw IllegalArgumentException("data.json not found in archive")
    }

    private fun deriveKey(password: String, salt: ByteArray): SecretKeySpec {
        val spec = PBEKeySpec(password.toCharArray(), salt, PBKDF2_ITERATIONS, 256)
        val bytes = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256").generateSecret(spec).encoded
        return SecretKeySpec(bytes, "AES")
    }

    fun exportEncrypted(context: Context, password: String): File {
        val payload = exportAllToJson(context).toString().toByteArray(Charsets.UTF_8)
        val salt = ByteArray(16).also { SecureRandom().nextBytes(it) }
        val nonce = ByteArray(12).also { SecureRandom().nextBytes(it) }
        val key = deriveKey(password, salt)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(128, nonce))
        val ciphertext = cipher.doFinal(payload)

        val file = File(backupsDir(context), timestampedName("backup_encrypted", "bak"))
        file.outputStream().use { out ->
            out.write(MAGIC.toByteArray())
            out.write(salt)
            out.write(nonce)
            out.write(ciphertext)
        }
        recordBackup(context, file.absolutePath, "encrypted")
        return file
    }

    fun importEncrypted(context: Context, file: File, password: String) {
        val bytes = file.readBytes()
        val magic = bytes.copyOfRange(0, 4).toString(Charsets.UTF_8)
        require(magic == MAGIC) { "Not a recognized encrypted backup file" }
        val salt = bytes.copyOfRange(4, 20)
        val nonce = bytes.copyOfRange(20, 32)
        val ciphertext = bytes.copyOfRange(32, bytes.size)
        val key = deriveKey(password, salt)

        val payload = try {
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(128, nonce))
            cipher.doFinal(ciphertext)
        } catch (e: Exception) {
            throw IllegalArgumentException("Incorrect password or corrupted backup file", e)
        }
        restoreFromJson(context, JSONObject(payload.toString(Charsets.UTF_8)))
    }
}
